pkg/controller/controller_multiplex.go

```diff
@@ -0,0 +1,175 @@
+
+const (
+	// ConditionReady represents that a given InstanceCondition is in
+	// ready state.
+	ConditionReady ConditionType = "Ready"
+
+	// ConditionFailed represents information about a final failure
+	// that should not be retried.
+	ConditionFailed ConditionType = "Failed"
+)
+
+// setCondition sets a single condition on a type's status. This delegates to
+// the correct type implementation.
+//
+// Note: objects coming from informers should never be mutated; always pass a
+// deep copy as the instance parameter.
+func setCondition(obj interface{},
+	condition ConditionType,
+	status v1beta1.ConditionStatus,
+	reason,
+	message string) {
+
+	switch toUpdate := obj.(type) {
+	case *v1beta1.ServiceInstance:
+		var conditionType v1beta1.ServiceInstanceConditionType
+		switch condition {
+		case ConditionReady:
+			conditionType = v1beta1.ServiceInstanceConditionReady
+		case ConditionFailed:
+			conditionType = v1beta1.ServiceInstanceConditionReady
+		}
+		setServiceInstanceCondition(toUpdate, conditionType, status, reason, message)
+		return
+	case *v1beta1.ServiceBinding:
+		var conditionType v1beta1.ServiceBindingConditionType
+		switch condition {
+		case ConditionReady:
+			conditionType = v1beta1.ServiceBindingConditionReady
+		case ConditionFailed:
+			conditionType = v1beta1.ServiceBindingConditionFailed
+		}
+		setServiceBindingCondition(toUpdate, conditionType, status, reason, message)
+		return
+	default:
+		glog.Errorf("attempting to set condition on object of unknown type: %+v", obj)
+	}
+}
+
+// Status
+
+// updateStatus inspects then updates status based on type
+func (c *controller) updateStatus(obj interface{}, pcb *pretty.ContextBuilder) (interface{}, error) {
+	switch toUpdate := obj.(type) {
+	case *v1beta1.ServiceInstance:
+		return c.updateServiceInstanceStatus(toUpdate)
+	case *v1beta1.ServiceBinding:
+		return c.updateServiceBindingStatus(toUpdate)
+	default:
+		s := fmt.Sprintf(
+			"attempting to set status on object of unknown type: %+v",
+			obj,
+		)
+		glog.Errorf(pcb.Message(s))
+		return obj, fmt.Errorf(s)
+	}
+}
+
+// prepareInProgressProperties generates the required properties for setting
+// the in-progress status of a Type
+func (c *controller) prepareInProgressProperties(object runtime.Object, toUpdate interface{}, namespace string, specParameters *runtime.RawExtension, specParametersFrom []v1beta1.ParametersFromSource, pcb *pretty.ContextBuilder) (map[string]interface{}, string, *runtime.RawExtension, error) {
+	var (
+		parameters                 map[string]interface{}
+		parametersChecksum         string
+		rawParametersWithRedaction *runtime.RawExtension
+		err                        error
+	)
+	if specParameters != nil || specParametersFrom != nil {
+		var parametersWithSecretsRedacted map[string]interface{}
+		parameters, parametersWithSecretsRedacted, err = buildParameters(c.kubeClient, namespace, specParametersFrom, specParameters)
+		if err != nil {
+			s := fmt.Sprintf(
```

Comment:
I am not convinced that it is worthwhile to do log, event, and status update in this function. I think that I would rather see the function return the `s` value --as an error--in the case of an error and have the caller use that to do the log, event, and status update. At that point, this function would be better place in parameters.go.